timeutil/schedule.go

```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+//
+//     timelist = timeset *( "," timeset )
+//     timeset = time / timespan
+//     time = 2DIGIT ":" 2DIGIT
+//     timespan = time ( "-" / "~" ) time [ "/" ( time / count ) ]
+//     count = 1*DIGIT
+//
+// Examples:
+// mon,10:00..fri,15:00 (Monday at 10:00, Friday at 15:00)
+// mon,fri,10:00,15:00 (Monday at 10:00 and 15:00, Friday at 10:00 and 15:00)
+// mon-wed,fri,9:00-11:00/2 (Monday to Wednesday and on Friday, twice between
+//                           9:00 and 11:00)
+// mon,9:00~11:00..wed,22:00~23:00 (Monday, sometime between 9:00 and 11:00, and
+//                                  on Wednesday, sometime between 22:00 and 23:00)
+// mon,wed  (Monday and on Wednesday)
+// mon..wed (same as above)
+//
+// Returns a slice of schedules or an error if parsing failed
+func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+	var schedule []*Schedule
+
+	for _, s := range strings.Split(scheduleSpec, "..") {
+		// cut the schedule in event sets
+		//     eventlist = eventset *( ".." eventset )
+		sched, err := parseEventSet(s)
+		if err != nil {
+			return nil, err
+		}
+		schedule = append(schedule, sched)
+	}
+	return schedule, nil
+}
+
+// parseWeekSpan parses a weekly span such as "mon-tue" or "mon2-tue3".
+func parseWeekSpan(s string) (span WeekSpan, err error) {
+	var parsed WeekSpan
+
+	split := strings.Split(s, spanToken)
+	if len(split) > 2 {
+		return span, fmt.Errorf("cannot parse %q: invalid week span", s)
+	}
+
+	parsed.Start, err = parseWeekday(split[0])
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: %q is not a valid weekday", s, split[0])
+	}
+
+	if len(split) == 2 {
+		parsed.End, err = parseWeekday(split[1])
+		if err != nil {
+			return span, fmt.Errorf("cannot parse %q: %q is not a valid weekday", s, split[1])
+		}
+	} else {
+		parsed.End = parsed.Start
+	}
+
+	if parsed.End.Pos < parsed.Start.Pos {
+		// eg. mon4-mon1
+		return span, fmt.Errorf("cannot parse %q: unsupported schedule", s)
+	}
+
+	if (parsed.Start.Pos != 0) != (parsed.End.Pos != 0) {
+		return span, fmt.Errorf("cannot parse %q: mixed weekday and nonweekday", s)
+	}
+
+	return parsed, nil
+}
+
+// parseClockSpan parses a time specification which can either be `<hh>:<mm>` or
+// `<hh>:<mm>[-~]<hh>:<mm>[/count]`. Alternatively the span can be one of
+// special tokens `-`, `~` (followed by an optional [/count]) that indicate a
+// whole day span, or a whole day span with spread respectively.
+func parseClockSpan(s string) (span ClockSpan, err error) {
+	var parsed ClockSpan
+	var rest string
+
+	// timespan = time ( "-" / "~" ) time [ "/" ( time / count ) ]
+
+	parsed.Split, rest, err = parseCount(s)
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: not a valid interval", s)
+	}
+
+	if strings.Contains(rest, spreadToken) {
+		// timespan uses "~" to indicate that the actual event
+		// time is to be spread.
+		parsed.Spread = true
+		rest = strings.Replace(rest,
+			spreadToken,
+			spanToken, 1)
```

Comment:
This should be a single line. We don't follow the 80 columns rule for Go in general, and this one in particular feels a bit over cramped.